README.md
## Notifier for Gitlab CI/CD Pipeline
[![Docker Cloud Build Status](https://img.shields.io/docker/cloud/build/ferrgo/gitlab-ci-notifier)](https://hub.docker.com/r/ferrgo/gitlab-ci-notifier)


[![Docker Pulls](https://img.shields.io/docker/pulls/ferrgo/gitlab-ci-notifier)](https://hub.docker.com/r/ferrgo/gitlab-ci-notifier)
[![Docker Image Size (latest semver)](https://img.shields.io/docker/image-size/ferrgo/gitlab-ci-notifier)](https://hub.docker.com/r/ferrgo/gitlab-ci-notifier)
[![Latest](https://img.shields.io/github/v/tag/ferrgo/gitlab-ci-notifier)](ferrgo/gitlab-ci-notifier/releases/latest)
[![License](https://img.shields.io/badge/license-GPL--3.0-blue.svg?style=flat)](LICENSE)

----
Shell script using [curl](https://curl.se/) and Gitlab CI Variables to post messages on Merge Requests

### Gitlab CI/CD Pipeline
See the [.gitlab-ci.yml](.gitlab-ci.yml) for the example and the [execution](https://gitlab.com/ferrgo/gitlab-ci-notifier/-/merge_requests/3)

### Usage
The script looks for the following variables:
- `TOKEN`: API Token, see [Personal access tokens](https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html#personal-access-tokens) or [Project access tokens](https://docs.gitlab.com/ee/user/project/settings/project_access_tokens.html)
- `CI_API_V4_URL`: API URL, always available on Gitlab Pipelines, see [Predefined variables reference](https://docs.gitlab.com/ee/ci/variables/predefined_variables.html#predefined-variables-reference)
- `CI_PROJECT_ID`: Unique id of the project, see [Predefined variables reference](https://docs.gitlab.com/ee/ci/variables/predefined_variables.html#predefined-variables-reference)
- `CI_MERGE_REQUEST_IID` or `CI_OPEN_MERGE_REQUESTS`: Variable for Merge Request ID, see [Predefined variables reference](https://docs.gitlab.com/ee/ci/variables/predefined_variables.html#predefined-variables-reference) or [Predefined variables for merge request pipelines](https://docs.gitlab.com/ee/ci/variables/predefined_variables.html#predefined-variables-for-merge-request-pipelines)
- `COMMENT_MSG`: Message to post on MR

#### Shell Usage
`cd src/ && ./notifier.sh`

#### Docker Usage
You can run using the docker image available on Docker Hub or GHCR

`docker run -e TOKEN=<TOkEN> -e CI_API_V4_URL=<URL> -e CI_PROJECT_ID=<ProjectID> -e CI_MERGE_REQUEST_IID=<MRID> -e COMMENT_MSG=<YourMessage> ferrgo/gitlab-ci-notifier`

### Docker image
Publishing to Docker Hub @ [ferrgo/gitlab-ci-notifier](https://hub.docker.com/r/ferrgo/gitlab-ci-notifier)

Publishing to Github Pkgs @ [docker.pkg.github.com/ferrgo/gitlab-ci-notifier/gitlab-ci-notifier](https://github.com/ferrgo/gitlab-ci-notifier/packages)

Publishing to Github Container Registry @ [ghcr.io/ferrgo/curlybox](https://github.com/users/ferrgo/packages/container/package/gitlab-ci-notifier)

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
- Script using cURL to post a message on a gitlab MR based on CI Variables
- Utillity logger shell functions
- README
- GPL v3 License
- Starting a Changelog file based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)

src/gitlab-commenter.sh
#!/bin/sh
. utils/logger.sh

comment_on_mr() {
  RES="$(post_curl)"
  if [ "${RES}" -ne 201 ]; then
    error "Curl failed to post message with code ""${RES}"" \nPayload:\n""$(cat .tmpcurl.log)""" 106
  fi
  log "Message posted \nHTTPCode:""${RES}"" \nPayload:\n""$(cat .tmpcurl.log)"""
  rm .tmpcurl.log
  log "Comment added on MR ""${MR_ID}"""
}

post_curl() {
  return $(curl \
    --request POST \
    --output .tmpcurl.log \
    --silent \
    --header "PRIVATE-TOKEN: ${TOKEN}" \
    --write-out "%{http_code}" \
    "${CI_API_V4_URL}/projects/${CI_PROJECT_ID}/merge_requests/${MR_ID}/notes" \
    --data "body=""${COMMENT_MSG}""")
}

set_mr_id() {
  if [ -n "${CI_OPEN_MERGE_REQUESTS}" ]; then
    log "Using list of open MR to get id"
    # List such as: 'gitlab-org/gitlab!333,gitlab-org/gitlab-foss!11'
    MR_ID=$(echo "${CI_OPEN_MERGE_REQUESTS}" | awk -F , '{print $1}' | awk -F ! '{print $2}')
  else
    log "Using current MR to get id"
    MR_ID=${CI_MERGE_REQUEST_IID}
  fi
}

check_variables() {
  if [ -z "${TOKEN}" ]; then error "API Token is not set or is blank" 101; fi
  if [ -z "${CI_API_V4_URL}" ]; then error "API URL is not set or blank" 102; fi
  if [ -z "${CI_PROJECT_ID}" ]; then error "ProjectID is not set or blank" 103; fi
  if [ -z "${CI_MERGE_REQUEST_IID}" ] && [ -z "${CI_OPEN_MERGE_REQUESTS}" ]; 
  then 
    error "No Variable for Merge Request ID" 104; 
  fi
  if [ -z "${COMMENT_MSG}" ]; then error "Message is not set or blank" 105; fi
}

main() {
  check_variables
  log "API: ${CI_API_V4_URL} ProjectID: ${CI_PROJECT_ID} mergeId: ${MR_ID} Message: ${COMMENT_MSG}"
  set_mr_id
  comment_on_mr
}

main

src/notifier.sh
#!/bin/sh
. utils/logger.sh

LOG_FILE=.tmpcurl.log

comment_on_mr() {
  RES=$(post_curl)
  if [ "${RES}" != "201" ]; then
    error 106 "Curl failed to post message with code ""${RES}"" " "Payload: ""$(cat ${LOG_FILE})"""
  fi
  log "Message posted" "HTTPCode:""${RES}"" " "Payload:" """$(cat ${LOG_FILE})"""
  rm ${LOG_FILE}
  log "Comment added on MR ""${MR_ID}"""
}

post_curl() {
  curl \
    --request POST \
    --output ${LOG_FILE} \
    -i \
    --header "PRIVATE-TOKEN: ${TOKEN}" \
    "${CI_API_V4_URL}/projects/${CI_PROJECT_ID}/merge_requests/${MR_ID}/notes" \
    --data "body=""${COMMENT_MSG}"""
  grep HTTP <${LOG_FILE} | awk '{print $2}'
}

set_mr_id() {
  if [ -n "${CI_OPEN_MERGE_REQUESTS}" ]; then
    log "Using list of open MR to get id"echo
    # List such as: 'gitlab-org/gitlab!333,gitlab-org/gitlab-foss!11'
    MR_ID=$(echo "${CI_OPEN_MERGE_REQUESTS}" | awk -F , '{print $1}' | awk -F ! '{print $2}')
  else
    log "Using current MR to get id"
    MR_ID=${CI_MERGE_REQUEST_IID}
  fi
}

check_variables() {
  if [ -z "${TOKEN}" ]; then error 101 "API Token is not set or is blank"; fi
  if [ -z "${CI_API_V4_URL}" ]; then error 102 "API URL is not set or blank"; fi
  if [ -z "${CI_PROJECT_ID}" ]; then error 103 "ProjectID is not set or blank"; fi
  if [ -z "${CI_MERGE_REQUEST_IID}" ] && [ -z "${CI_OPEN_MERGE_REQUESTS}" ]; then
    error 104 "No Variable for Merge Request ID"
  fi
  if [ -z "${COMMENT_MSG}" ]; then error 105 "Message is not set or blank"; fi
}

main() {
  check_variables
  set_mr_id
  log "API: ${CI_API_V4_URL} ProjectID: ${CI_PROJECT_ID} mergeId: ${MR_ID} Message: ${COMMENT_MSG}"
  comment_on_mr
}

main
